frame/staking/src/offchain_election.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helpers for offchain worker election.
+
+use crate::{Call, CompactAssignments, Module, SessionInterface, Trait};
+use codec::Encode;
+use frame_system::offchain::{SubmitUnsignedTransaction};
+use sp_phragmen::{reduce, ExtendedBalance, PhragmenResult, StakedAssignment};
+use sp_std::{prelude::*, cmp::Ordering};
+use sp_runtime::{RuntimeAppPublic, RuntimeDebug};
+
+#[derive(RuntimeDebug)]
+pub(crate) enum OffchainElectionError {
+	/// No signing key has been found on the current node that maps to a validators. This node
+	/// should not run the offchain election code.
+	NoSigningKey,
+	/// Phragmen election returned None. This means less candidate that minimum number of needed
+	/// validators were present. The chain is in trouble and not much that we can do about it.
+	FailedElection,
+}
+
+/// Compares two sets of phragmen scores based on desirability and returns true if `that` is
+/// better `this`.
+///
+/// Evaluation is done in a lexicographic manner.
+pub(crate) fn is_score_better(this: [ExtendedBalance; 3], that: [ExtendedBalance; 3]) -> bool {
+	match that
+		.iter()
+		.enumerate()
+		.map(|(i, e)| e.cmp(&this[i]))
+		.collect::<Vec<Ordering>>()
+		.as_slice()
+	{
+		[Ordering::Greater, _, _] => true,
+		[Ordering::Equal, Ordering::Greater, _] => true,
+		[Ordering::Equal, Ordering::Equal, Ordering::Less] => true,
+		_ => false,
+	}
+}
+
+/// The internal logic of the offchain worker of this module.
+pub(crate) fn compute_offchain_election<T: Trait>() -> Result<(), OffchainElectionError> {
+	let validator_keys = T::SessionInterface::keys::<T::KeyType>();
+	let local_keys = T::KeyType::all();
+
+	if let Some((index, ref pubkey)) = local_keys
```

Comment:
There might be multiple keys on a single node - this code should run in a loop.

Reply:
I actually intentionally did the opposite. 

My reasoning was: If keystore is sane, you will have one instance of each key type IFIU. 
If for any reason this is not the case, then lets not propagate the issue further and just use the first key that matches.

Reply:
We had an idea floating around some time ago that it should be possible to have multiple identities on a single node (i.e. you have 3 validators, but they all pretty much run on the same node). That's why there might be multiple keys in a keystore.
However if that idea is fully implemented and currently possible, but still I feel like a loop is just the right approach here (you might want to insert a new key and deprecate the old one, so temporarily you have two keys, and sending from one of them might fail).

Reply:
>  (you might want to insert a new key and deprecate the old one, so temporarily you have two keys, and sending from one of them might fail).

Still have my doubts tbh. my worry is that you might really f* this up and have a wrong keystore and send from like.. a hundred wrong keys. Not sending is way safer in this case, given that you might end up spamming  Let's leave this open. I will consider it and it won't be a big code change anyhow.

Reply:
>  send from like.. a hundred wrong keys. 

How is that even possible? You intersect a list of local accounts with the validator set, it's not really possible that you have a bunch of keys in the keystore that somehow are in the validator set as well, but are invalid (?) and you end up spamming.

Or are you trying to say that since the transaction is going to be the same for all validators (just differing in signature) it doesn't make sense to submit it multiple times? That definitely makes sense, but it should probably be a loop anyway and you should break out of it on first successful submission to the pool.

Reply:
> Or are you trying to say that since the transaction is going to be the same for all validators (just differing in signature) it doesn't make sense to submit it multiple times?

I didn't think of it before but yeah that's also a point. 

> , but it should probably be a loop anyway and you should break out of it on first successful submission to the pool.

This sounds like best of both opinions. Lets go with this.

Reply:
done.